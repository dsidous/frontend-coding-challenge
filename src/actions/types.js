export const types = {
  FETCH_TOURNAMENTS: 'fetchTournaments',
  FETCH_TOURNAMENTS_SUCCESS: 'fetchTournamentsSuccess',
  FETCH_TOURNAMENTS_ERROR: 'fetchTournamentsError',

  UPDATE_TOURNAMENT: 'updateTournamentError',
  UPDATE_TOURNAMENT_ERROR: 'updateTournamentError',

  DELETE_TOURNAMENT: 'deleteTournament',
  DELETE_TOURNAMENT_ERROR: 'deleteTournamentError',

  CREATE_TOURNAMENT: 'createTournament',

  EDIT_TOURNAMENT: 'editTournament',

  DELETE_TOURNAMENT: 'deleteTournament',

  QUERY: 'query'
};
