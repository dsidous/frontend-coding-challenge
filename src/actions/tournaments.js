import { types } from './types';
import axios from 'axios';
import { API_TOURNAMENTS_URL } from '../constants/api';

export const fetchTournaments = () => async (dispatch, getState) => {
  const query = getState().query.query || '';

  try {
    dispatch({
      type: types.FETCH_TOURNAMENTS
    });

    const response = await axios.get(`${API_TOURNAMENTS_URL}?q=${query}`);
    const data = await response.data;

    dispatch({
      type: types.FETCH_TOURNAMENTS_SUCCESS,
      payload: data
    });
  } catch (error) {
    dispatch({
      type: types.FETCH_TOURNAMENTS_ERROR
    });
  }
};

export const createTournament = newName => async dispatch => {
  try {
    await axios.post(`${API_TOURNAMENTS_URL}`, { name: newName });
    dispatch(fetchTournaments());
  } catch (error) {
    dispatch({
      type: types.CREATE_TOURNAMENT_ERROR
    });
  }
};

export const updateTournament = (id, newName) => async dispatch => {
  try {
    await axios.patch(`${API_TOURNAMENTS_URL}/${id}`, { name: newName });
    dispatch(fetchTournaments());
  } catch (error) {
    dispatch({
      type: types.UPDATE_TOURNAMENT_ERROR
    });
  }
};

export const deleteTournament = id => async dispatch => {
  try {
    await axios.delete(`${API_TOURNAMENTS_URL}/${id}`);
    dispatch(fetchTournaments());
  } catch (error) {
    dispatch({
      type: types.DELETE_TOURNAMENT_ERROR
    });
  }
};
