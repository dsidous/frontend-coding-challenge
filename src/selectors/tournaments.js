export const getTournaments = state => {
  return state;
};
