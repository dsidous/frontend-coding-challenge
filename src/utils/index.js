import configureStore from 'redux-mock-store';
import { applyMiddleware, createStore } from 'redux';
import rootReducer from '../reducers';
import { middlewares } from '../store';

export const testStoreState = initialState => {
  const createStoreWithMiddleware = applyMiddleware(...middlewares)(
    createStore
  );
  return createStoreWithMiddleware(rootReducer, initialState);
};

export const testStore = configureStore(middlewares);
