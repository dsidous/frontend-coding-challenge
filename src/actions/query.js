import { types } from './types';
import { fetchTournaments } from './tournaments';

export const addQuery = query => async dispatch => {
  try {
    dispatch({
      type: types.QUERY,
      payload: { query }
    });
    dispatch(fetchTournaments());
  } catch (error) {
    console.log(error);
  }
};
