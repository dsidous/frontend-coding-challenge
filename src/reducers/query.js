import { types } from '../actions/types';
const initialState = [];

export default function query(state = initialState, { type, payload }) {
  switch (type) {
    case types.QUERY:
      return {
        ...state,
        query: payload.query
      };
    default:
      return state;
  }
}
