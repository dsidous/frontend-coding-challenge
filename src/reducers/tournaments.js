import { types } from '../actions/types';
const initialState = [];

export default function tournaments(state = initialState, { type, payload }) {
  switch (type) {
    case types.FETCH_TOURNAMENTS:
      return {
        ...state,
        loading: true,
        error: false
      };
    case types.FETCH_TOURNAMENTS_ERROR:
      return {
        ...state,
        loading: false,
        error: true
      };
    case types.FETCH_TOURNAMENTS_SUCCESS:
      return {
        ...state,
        tournaments: payload,
        loading: false,
        error: false
      };
    default:
      return state;
  }
}
