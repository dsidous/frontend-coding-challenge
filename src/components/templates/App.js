import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import styled from 'styled-components';
import { fetchTournaments, createTournament } from '../../actions/tournaments';
import { addQuery } from '../../actions/query';

import Container from '../atoms/Container';
import H4 from '../atoms/H4';
import Input from '../atoms/Input';
import Button from '../atoms/Button';
import TournamentList from '../organisms/TournamentList';
import { getTournaments } from '../../reducers';

import theme from '../../theme';

const StyledContainer = styled.div`
  display: flex;
  justify-content: space-between;
  margin-bottom: ${theme.spacing(3)};
`;

const App = () => {
  const [query, setQuery] = useState('');

  const data = useSelector(state => getTournaments(state));
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchTournaments());
  }, [dispatch]);

  const handleSubmit = e => {
    e.preventDefault();
    dispatch(addQuery(query));
  };

  const newTournament = () => {
    const newName = window.prompt('Tournament Name:');
    if (newName) {
      dispatch(createTournament(newName));
    }
  };

  return (
    <Container>
      <H4>FACEIT Tournaments</H4>
      <Container>
        <StyledContainer>
          <form onSubmit={handleSubmit} data-test="searchForm">
            <Input
              data-test="searchInput"
              placeholder="Search tournament..."
              onChange={e => setQuery(e.target.value)}
              value={query}
            />
          </form>
          <Button data-test="buttonCreate" onClick={newTournament}>
            Create Tournament
          </Button>
        </StyledContainer>
        <TournamentList data={data} />
      </Container>
    </Container>
  );
};

export default App;
