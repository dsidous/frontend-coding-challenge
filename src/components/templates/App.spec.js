import React from 'react';
import { shallow, mount } from 'enzyme';

import App from './App';
import * as tournament from '../../actions/tournaments';
import * as query from '../../actions/query';
import { findByTestAtrr } from '../../utils';

tournament.fetchTournament = jest.fn();
tournament.createTournament = jest.fn();

query.addQuery = jest.fn();

const mockProps = {
  tournament: {
    id: 'mock id',
    name: 'mock name',
    organizer: 'mock organizer',
    game: 'mock game',
    participants: {
      current: 1,
      max: 2
    },
    startDate: '2020.12.12 10:10:10'
  }
};

const mockDispatch = jest.fn();

jest.mock('react-redux', () => ({
  useSelector: jest.fn(() => ({ ...mockProps })),
  useDispatch: () => jest.fn()
}));

describe('templates/App', () => {
  let wrapper;

  it('Should render as expected', () => {
    wrapper = shallow(<App />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('On Create new tournament', () => {
    let promptSpy;
    let createBtn;
    const newName = 'New Mock Name';

    beforeEach(() => {
      wrapper = shallow(<App />);
      createBtn = findByTestAtrr(wrapper, 'buttonCreate');
      promptSpy = jest.spyOn(window, 'prompt');
      promptSpy.mockImplementation(jest.fn(() => newName));
      jest.spyOn(tournament, 'createTournament');
    });

    afterEach(() => {
      promptSpy.mockRestore();
      jest.clearAllMocks();
    });

    it('Should call the prompt window', () => {
      createBtn.simulate('click');
      expect(promptSpy).toBeCalled();
    });

    it('Should dispatch action with the new name if clicked OK in the prompt window', () => {
      createBtn.simulate('click');
      expect(tournament.createTournament).toBeCalledWith(newName);
    });

    it("Shouldn't dispatch action if clicked Cancel in the prompt window", () => {
      promptSpy.mockImplementation(jest.fn(() => false));

      createBtn.simulate('click');
      expect(mockDispatch).not.toBeCalled();
    });

    describe('On Search', () => {
      it('Should dispath action on submit with the query', async () => {
        wrapper = mount(<App />);
        const querySpy = jest.spyOn(query, 'addQuery');

        const queryMock = 'mock query';
        const input = findByTestAtrr(wrapper, 'searchInput').find('input');
        const form = findByTestAtrr(wrapper, 'searchForm');

        input.instance().value = queryMock;
        input.simulate('change', { target: { value: queryMock } });
        form.simulate('submit', { preventDefault: jest.fn() });

        expect(querySpy).toBeCalledWith(queryMock);

        querySpy.mockRestore();
      });
    });
  });
});
