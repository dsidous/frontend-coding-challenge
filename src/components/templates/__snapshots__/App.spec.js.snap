// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templates/App Should render as expected 1`] = `ShallowWrapper {}`;
